File: src/main/java/org/kite9/framework/dom/AbstractDiagramElement.java
package org.kite9.framework.dom;

import org.kite9.diagram.common.HintMap;
import org.kite9.diagram.model.Diagram;
import org.kite9.diagram.model.DiagramElement;

public abstract class AbstractDiagramElement implements DiagramElement {

	protected DiagramElement parent;
	protected HintMap hints;
	
	public AbstractDiagramElement(DiagramElement parent) {
		super();
		this.parent = parent;
	}

	public int compareTo(DiagramElement o) {
		return getID().compareTo(o.getID());
	}

	@Override
	public int hashCode() {
		String id = getID();
		return id.hashCode();
	}

	public AbstractDiagramElement() {
		super();
	}

	@Override
	public DiagramElement getParent() {
		return parent;
	}
	
	private int depth = -1;

	@Override
	public int getDepth() {
		if (depth == -1) {
			if (this instanceof Diagram) {
				depth = 0;
			} else {
				depth = getParent().getDepth() + 1;
			}
		}

		return depth;
	}
	
	

}

File: kite9-visualization/src/common/kotlin/org/kite9/diagram/common/elements/factory/AbstractDiagramElement.kt
package org.kite9.diagram.common.elements.factory

import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.Diagram
import org.kite9.diagram.model.DiagramElement

abstract class AbstractDiagramElement(private val p: DiagramElement?) : DiagramElement {

    override fun compareTo(other: DiagramElement): Int {
        return getID().compareTo(other.getID())
    }

    override fun hashCode(): Int {
        return _hc;
    }

    private var depth = -1

    private val _hc : Int by lazy { getID().hashCode() }

    override fun getDepth(): Int {
        if (depth == -1) {
            depth = if ((this is Diagram) || (getParent() == null)){
                0
            } else {
                getParent()!!.getDepth()  +1
            }
        }
        return depth
    }

    /**
     * Used to keep track of any connections we find in the diagram during initialisation
     */
    protected open fun registerConnection(de: Connection) {
        (getParent() as AbstractDiagramElement).registerConnection(de)
    }

    override fun getParent(): DiagramElement? {
        return p
    }
}